File: EducationalScripts/Example2.py
import sys
from tempfile import TemporaryFile
from org.apache.commons.io import IOUtils
from java.nio.charset import StandardCharsets
from org.apache.nifi.processor.io import StreamCallback

reload(sys)
sys.setdefaultencoding('UTF-8')
additionalContent = "test"

class ModifyContent(StreamCallback):
    def __init__(self):
        pass

    def process(self, inputStream, outputStream):
        global additionalContent
        flowFileContent = IOUtils.toString(inputStream, StandardCharsets.UTF_8)
        flowFileContent = flowFileContent + "\n" + additionalContent
        outputStream.write(bytearray(flowFileContent.encode('utf-8')))       

flowFile = session.get()

if flowFile is not None:
    try:
        session.write(flowFile, ModifyContent())
    except Exception as e:
        flowFile = session.putAttribute(flowFile, "error", str(e))
        session.transfer(flowFile, REL_FAILURE)
    else:
        session.transfer(flowFile, REL_SUCCESS)
    finally:
        session.commit()

File: EducationalScripts/Example2.clj
(import 'java.nio.charset.StandardCharsets)
(import 'org.apache.commons.io.IOUtils)
(import 'org.apache.nifi.processor.io.StreamCallback)

;; (def additionalContent "test")
(def newAttrs {"newAttr1" "apple"
               "newAttr2" (str 12)})

(defn stream-callback [flow content]
  (reify StreamCallback
    (process [this inputStream outputStream]
      (.write outputStream
              (.getBytes (str (slurp inputStream) "\n" content))))))

(defn write-content[flow content]
  (.write session flow (stream-callback flow content)))

(let [flowFile (.get session)]
  (try
  (def flowFileAttrs (.getAttributes flowFile))
  (write-content flowFile flowFileAttrs)
  (.putAllAttributes session flowFile newAttrs)
  (.putAttribute session flowFile "modify" "true") 
  (.info log "Content has been modified!")
  (.transfer session flowFile REL_SUCCESS)
    (catch Exception e
      (.putAttribute session flowFile "error" (.getMessage e))
      (.transfer session flowFile REL_FAILURE))
    (finally (.commit session))))
